File: code/python/10.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        if not p: return not s
        if not s and len(p) == 1: return False
        print(p)
        nrow = len(s) + 1
        ncol = len(p) + 1

        dp = [[False for c in range(ncol)] for r in range(nrow)]

        dp[0][0] = True
        dp[0][1] = False
        for c in range(2, ncol):
            j = c - 1
            # 这里从dp[0][0]开始的话，是将*当作0个来用的
            if p[j] == '*': dp[0][c] = dp[0][c - 2]

        for r in range(1, nrow):
            i = r - 1
            for c in range(1, ncol):
                j = c - 1
                if s[i] == p[j] or p[j] == '.':
                    dp[r][c] = dp[r - 1][c - 1]
                elif p[j] == '*':
                    if p[j - 1] == s[i] or p[j - 1] == '.':
                        dp[r][c] = dp[r - 1][c] or dp[r][c - 2]
                    else:
                        dp[r][c] = dp[r][c - 2]
                else:
                    dp[r][c] = False

        return dp[nrow - 1][ncol - 1]

File: code/java/10.java
class Solution {
    public boolean isMatch(String s, String p) {
        int row = s.length();
        int col = p.length();
        boolean[][] dp =new boolean[row+1][col+1];

        dp[0][0] = true;
        for(int i=2;i<=col;i++){
            if(p.charAt(i-1)=='*') dp[0][i] = dp[0][i-2];
        }

        for(int i=1;i<=row;i++){
            int r = i-1;
            for(int j=1;j<=col;j++){
                int c = j-1;
                if(p.charAt(c)=='*'){
                    if(p.charAt(c-1)==s.charAt(r)||p.charAt(c-1)=='.')
                        dp[i][j] = dp[i-1][j]||dp[i][j-2];
                    else
                        dp[i][j] = dp[i][j-2];
                }else{
                    if(p.charAt(c)=='.'||s.charAt(r)==p.charAt(c)) dp[i][j] = dp[i-1][j-1];
                    else dp[i][j] = false;
                }
            }
        }
        return dp[row][col];
    }
}